### PR TITLE
Upgrade spack version used by mason

### LINKED ISSUE #19517: mason test failures
### Summary of Problem

When I run `start_test mason`, I get three failures:

```
[Error matching program output for mason/env/mason-registry]
[Error matching program output for mason/mason-external/libtomlc99/mason-external]
[Error matching program output for mason/mason-external/masonExternalRanges/mason-external-range]
[Summary: #Successes = 88 | #Failures = 3 | #Futures = 0 | #Warnings = 0 ]
[Summary: #Passing Suppressions = 0 | #Passing Futures = 0 ]
```

The first looks like a git thing.  The third is a python-looking error, maybe from spack?  I don't know about the second one.

```
[Executing diff mason-registry.good mason-registry.exec.out.tmp]
7a8,17
> hint: Using 'master' as the name for the initial branch. This default branch name
> hint: is subject to change. To configure the initial branch name to use in all
> hint: of your new repositories, which will suppress this warning, call:
> hint:
> hint:         git config --global init.defaultBranch <name>
> hint:
> hint: Names commonly chosen instead of 'master' are 'main', 'trunk' and
> hint: 'development'. The just-created branch can be renamed via this command:
> hint:
> hint:         git branch -m <name>
```

```
[Executing diff mason-external.good mason-external.exec.out.tmp]
2,5c2
< Skipping registry update since no dependency found in manifest file.
< Compiling [debug] target: CToml
< Build Successful
<
---
> Package could not be installed
```

```
[Executing diff mason-external-range.good mason-external-range.exec.out.tmp]
2,5c2,6
< Skipping registry update since no dependency found in manifest file.
< Compiling [debug] target: masonExternalRanges
< Build Successful
<
---
> ==> Error: Failed to install libtomlc99 due to AttributeError: 'NoneType' object has no attribute 'replace'
> ==> Error: 'NoneType' object has no attribute 'replace'
> uncaught MasonError: Package could not be installed
>   $CHPL_HOME/tools/mason//MasonExternal.chpl:606: thrown here
>   mason-external-range.chpl:14: uncaught here
[Error matching program output for mason/mason-external/masonExternalRanges/mason-external-range]
```

### Configuration Information

```
chpl version 1.26.0

CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: gnu
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: native
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none
CHPL_TASKS: qthreads
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
CHPL_GMP: none
CHPL_HWLOC: bundled
CHPL_RE2: bundled
CHPL_LLVM: none *
CHPL_AUX_FILESYS: none

gcc (Ubuntu 11.2.0-7ubuntu2) 11.2.0

git version 2.32.0

Python 3.9.7

/etc/lsb-release
DISTRIB_ID=Pop
DISTRIB_RELEASE=21.10
DISTRIB_CODENAME=impish
DISTRIB_DESCRIPTION="Pop!_OS 21.10"
```

`test/mason/mason-external/masonExternalRanges/mason_home/spack` ended up with 0.15.4 checked out.

### PR DESCRIPTION
**Pull request by ronawho**
We've been using spack 0.15.4 for mason-external for a while because newer versions didn't work when spack was in a path longer than ~100 chars. However, not upgrading has left us with a spack that doesn't work on M1 macs or python >= 3.8, which seems worse to me.

This upgrades spack to 0.19.0, which works on M1 and supports python 2.7/3.6-3.11. It also updates the mason-external tests to use a `MASON_HOME`/`SPACK_ROOT` in `CHPL_TEST_TMP_DIR`, which generally has a short directory structure so should avoid the path length issues. I'm using `CHPL_TEST_TMP_DIR` instead of just a new tmpdir because we take extra care to clean up `CHPL_TEST_TMP_DIR`.

A side benefit of this is that that dir is usually in `/tmp` so on some systems we avoid cloning spack into NFS/lustre so this should save some time and reduce load on shared filesystems.

Testing:
 - python 3.10 on x86 linux with short and long paths
 - python 3.7 on x86 linux
 - python 3.9 on aarch mac

Resolves #19517
Resolves Cray/chapel-private#3206
Resolves Cray/chapel-private#3518
Part of Cray/chapel-private#4110

### CODE REVIEW

**Comment by ronawho** [comment]
@arezaii @bmcdonald3 I have tested this with python 3.10 in both a chapel dir that's very short and very long (that causes test failures if we didn't set MASON_HOME to a tmp dir) so I feel pretty confident this will work for our testing (unless we have nightly configs with long `CHPL_TEST_TMP_DIR` paths)

This does mean users could get bitten by the long path issue, but I think that's better than not working with new pythons or M1s and there are workarounds so overall I feel positive about the change.

**Review by bmcdonald3 — APPROVED** [review]
Thanks Elliot, looks good to me.

File: tools/mason/MasonExternal.chpl [hunk 1]
```diff
@@ -261,10 +261,10 @@ proc getSpackVersion : VersionInfo {
   const command = "spack --version";
   const tmpVersion = getSpackResult(command,true).strip();
   // on systems with their own spack, spack --version can provide
-  // a version string like x.x.x-xxxx-hash
+  // a version string like x.x.x.xxxx (hash)
   // partitioning the string allows us to separate the major.minor.bug
   // from the remaining values
-  const version = tmpVersion.partition("-");
+  const version = tmpVersion.partition(" ");
```
Comment (mppf) [inline tools/mason/MasonExternal.chpl:267]: Elliot I see this is the same change as the one I did in https://github.com/chapel-lang/chapel/pull/21439/files but I think my version allows both formats. Do you think that we should allow both version formats here or let the old format go away & say that `mason external` only supports new enough spack versions?

Comment (ronawho) [reply tools/mason/MasonExternal.chpl:267]: In your PR, I think `nodash` was unused, but in general I'm good saying we only support new enough spack versions.

**Review by arezaii — APPROVED** [review]

**Pull request merged**